draft-ietf-quic-transport.md

```diff
@@ -2194,11 +2194,13 @@ linked by any other entity.
 At any time, endpoints MAY change the Destination Connection ID they send to a
 value that has not been used on another path.
 
-An endpoint MUST use a new connection ID if it initiates connection migration as
-described in {{initiating-migration}} or probes a new network path as described
-in {{probing}}.  An endpoint MUST use a new connection ID in response to a
-change in the address of a peer if the packet with the new peer address uses an
-active connection ID that has not been previously used by the peer.
+An endpoint MUST use a new connection ID when it initiates connection migration
+as described in {{initiating-migration}} or {{preferred-address}}, or when it
+probes a new network path as described in {{probing}}, and MUST NOT reuse
```

Comment:
The run-on sentence here could be two sentences.

```suggestion
probes a new network path as described in {{probing}}. Endpoints MUST NOT reuse
```

...but I'm not seeing what the new content here does.  It appears to just be a restatement of the previous clause.

Reply:
The problem with the existing text is that while it says that a new connection ID MUST be used when "initiating" a migration, or probing, it does not forbid reuse of a CID (that has previously used on a different path) once the path is being validated.

That's is why I added the "MUST NOT" without splitting the sentence.

Reply:
The "MUST NOT" stands on its own in that case:

> An endpoint MUST NOT use a connection ID on a given network path if it has used the same connection ID to send packets on another network path.

The intended meaning of "new" in this text is to imply "previously unused", which might be more clearly stated as:

> An endpoint MUST use a connection ID that it has not previously used when it initiates connection migration ({{initiating-migration}}), including using a preferred address ({{preferred-address}}).  Similarly, an endpoint MUST use an unused connection when probing a new network path ({{probing}}).   An endpoint MUST use a previously-unused connection ID in response to a change in the address of a peer if the packet with the new peer address uses an active connection ID that has not been previously used by the peer.  An endpoint MUST NOT [...as above].

This covers all the existing reasons to use a previously unused connection ID.  You could restate as a list instead, I guess:

> An endpoint MUST use a new connection ID when it:
> * migrates to a new network path ({{initiating-migration}}), including migrating to a server's preferred address ({{preferred-address}});
> * probes a new network ({{probing}}); or,
> * sends a packet on a new network path as a result of responding to the migration of a peer, where the peer uses a connection ID that has not been used on another path.
>In all cases, an endpoint MUST NOT use a connection ID on a given network path if it has used the same connection ID to send packets on another network path.

Reply:
We do have this text in [5.1.2](https://quicwg.org/base-drafts/draft-ietf-quic-transport.html#name-consuming-and-retiring-conn):
> As discussed in Section 9.5, each connection ID MUST be used on packets sent from only one local address. An endpoint that migrates away from a local address SHOULD retire all connection IDs used on that address once it no longer plans to use that address.

Is the missing piece that it should also be used for only one remote address, but this is the only scenario in which the server changes address?

Reply:
Yes. IIUC, the requirements are:
* A client MUST NOT reuse a CID across multiple paths (being identified by the 5 tuple).
* A server MUST NOT reuse a CID across multiple path, with the exception being that it MAY use a CID that has been used on one path on another path, if that new path is opened by a packet sent by a client using DCID for the old path.

We can combine these statements into one, and say that: An endpoint MUST NOT reuse a CID across multiple paths, with the exception being that it MAY use a CID that has been used on one path on another path, if that new path is opened by a packet sent by a client using DCID for the old path.

I think that is the most precise definition I can think of. If the above sounds fine, I can work on it.

Reply:
Writing down the results of a discussion in ZRH about stating this in terms of endpoints (not client/server) and addresses (not paths):
- An endpoint MUST NOT reuse a CID when sending from more than one local address
- An endpoint MUST NOT reuse a CID when sending to more than one destination address

With the following exceptions/additions:
- An endpoint SHOULD NOT respond to a newly seen incoming CID with an outgoing CID that it has used before (should since it may not have a new one to use)
- An endpoint MAY skip switching if it has strong reason to believe that the remote endpoint didn't try to change (i.e. if CID didn't change and only port changed, it's likely NAT rebinding), although this exception is only necessary if it doesn't have a new one to use

(Modulo some wording tweaks)

I *think* this is the current state of the text, just rearranging to collapse the separate requirements into one cluster that's hopefully easy to understand. Will incorporate this into the SPA PR if that seems reasonable to you.

Reply:
@erickinnear I think "MUST NOT" should be "MUST NOT deliberately" because NAT rebinding.